docs/design.md

```diff
@@ -0,0 +1,279 @@
+for data providers to be able to interact with indexer nodes. 
+
+## Overview
+- Data providers keep a local index of all the data they are providing.
+- Data providers are responsible for the advertisement of updates related to the data they are providing.
+        - Data providers must be able to uniquely identify the batches of data they are providing, and/or the
+        announcements they broadcast to the network in order to be able to serve indexer-node data ingestion requests.
+- Indexer nodes listen to advertisement from data providers. When they come across a previously unseeend
+advertisement for a data provider, they trigger a new ingestion session to sync with the provider. 
+
+## Data Structures
+Data providers and indexer nodes can use any internal representation they want for indexed data. They only
+data structures enforced by the protocol are in the interaction between the two.
+
+### Advertisements
+Advertisements are used by data providers to share updates about the data they are providing. Ideally,
+these advertisements are published in a shared pub-sub channel between providers and indexer nodes. However,
+some indexer nodes may choose to provide a dedicated endpoint to allow certain authenticated providers to push announcements and updates directly to them.
+- `Pull` Advertisements: Indexer-nodes pull advertisement from a shared pub-sub channel.
+- `Push` Advertisements: Authenticated providers can push advertisements to indexer-nodes.
+
+```go
+// AdvertisementID uniquely identifies an advertisement.
+// It is the multihash of IndexID + Provider + Previous.
+type AdvertisementID mh.Multihash
+
+type Advertisement struct {
+        ID              AdvertisementID
+        IndexID         cid.Cid
+        Previous        AdvertisementID
+        Provider        peer.ID
+        Signature       []byte
+        GraphSupport    bool
+        
+}
+```
+- `ID` is the unique identifier used by the data provider to identify this update.
+- `IndexID` is the CID of the data included in the Advertisement. It can be used by the indexer to authenticate the
+advertisement and the data ingested from the provider.
+- `Previous` links the current advertisement for the provider with the previous one. This allows
+providers to track the announcements they are broadcasting, and 
+indexer nodes to realize when they have missed a previous advertisement.
+- `Provider` determines the source of the advertisement.
+- `Signature` is a signature over `Advertisement.ID` to authenticate that the advertisement has been generated by `Provider`.
+- `GraphSupport` is used by providers to signal indexer nodes that they support IPLD-aware ingestion sessions (see [protocol](#protocol)).
+
+Indexer nodes use these IDs in their ingestion requests to fetch indexing
+data from the provider.
+
+For instance, a Filecoin miner may include as part of a new `Advertisement` the set of CIDs included in a new deal, and it may
```

Comment:
I do not think the Advertisement should ever include CIDs.  This should only be something that is fetched by the indexer.  There are a number of reasons for this:
- Too much data for pubsub
- Need response pagination
- Indexer should get data only when it determines it is time for an update
- Allows for batching and for skipped updates

Reply:
This is a misunderstanding, I don't expect Advertisements to include all CIDs for a deal, but to reference a specific set of CIDs. It is probably not described correctly, I'll rephrase it.